File: src/ObjectManager/Object.Tes/FilePacks/BsaMultiFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace OA.Tes.FilePacks
{
    public partial class BsaMultiFile : IDisposable
    {
        public readonly List<BsaFile> Packs = new List<BsaFile>();

        public BsaMultiFile(string[] filePaths)
        {
            if (filePaths == null)
                throw new ArgumentNullException("filePaths");
            var files = filePaths.Where(x => Path.GetExtension(x) == ".bsa" || Path.GetExtension(x) == ".ba2");
            Packs.AddRange(files.Select(x => new BsaFile(x)));
        }

        void IDisposable.Dispose()
        {
            Close();
        }

        ~BsaMultiFile()
        {
            Close();
        }

        public void Close()
        {
            foreach (var pack in Packs)
                pack.Close();
        }

        /// <summary>
        /// Determines whether the BSA archive contains a file.
        /// </summary>
        public virtual bool ContainsFile(string filePath)
        {
            return Packs.Any(x => x.ContainsFile(filePath));
        }

        /// <summary>
        /// Loads an archived file's data.
        /// </summary>
        public virtual byte[] LoadFileData(string filePath)
        {
            var pack = Packs.FirstOrDefault(x => x.ContainsFile(filePath));
            if (pack == null)
                throw new FileNotFoundException($"Could not find file \"{filePath}\" in a BSA file.");
            return pack.LoadFileData(filePath);
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/BsaMultiFile.swift
//
//  BsaMultiFile.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation

public class BsaMultiFile {
    public var packs = [BsaFile]()

    init(_ filePaths: [URL]?) {
        guard let filePaths = filePaths else {
            fatalError("Requires filePaths")
        }
        filePaths.filter { ["bsa", "ba2"].index(of: $0.pathExtension.lowercased()) != nil }
            .map { BsaFile($0.path) }
            .forEach { packs.append($0) }        
//        let file1 = loadFileData("trees\\treecottonwoodsu.spt")
//        debugPrint("file1", file1)
    }

    deinit {
        close()
    }

    public func close() {
        for pack in packs {
            pack.close()
        }
    }

    public func containsFile(_ filePath: String) -> Bool {
        return packs.contains { $0.containsFile(filePath) }
    }

    public func loadFileData(_ filePath: String) -> Data {
        guard let pack = packs.first(where: { $0.containsFile(filePath) }) else {
            fatalError("Could not find file '\(filePath)' in a BSA file.")
        }
        return pack.loadFileData(filePath: filePath)
    }
}
